Fock representation of Symplectic transformations

nquesada opened the pull request:
**Context:**
It is desirable to have a way to construct the Fock representation of Gaussian transformation 
**Description of the Change:**
It adds the module operation to obtain such functionality
**Benefits:**
One can now construct tensors representing Gaussian operations

nquesada commented:
The name `operations` is very tentative. I find it very bland. Suggestions naming suggestions are also welcomed!

josh146 commented:
Since the module contains Fock representations of Gaussian operations, I quite like just `fock.py`.

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+"""
+
+from itertools import product
+import numpy as np
```
```suggestion

import numpy as np

```

nquesada commented:
`from thewalrus import fock`

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+
+def n_two_mode_squeezed_vac(n, r=np.arcsinh(1.0)):
+    r"""
+    Returns the symplectic matrix associated with two mode squeezing operations by amount r between modes i and i+n for 0<=i<n
```
Better to move the details regarding `n` and `r` into the body of the docstring, or into the args list
```suggestion
    Returns the symplectic matrix associated with two mode squeezing operations.
```

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+from thewalrus.quantum import Amat
+
+
+def n_two_mode_squeezed_vac(n, r=np.arcsinh(1.0)):
```
Two questions:

1. Isn't S2gate already in thewalrus.symplectic?

2. This symplectic transformation can be applied to any state, so why does the name have `vac`?

nquesada replied:
1. Yes. This is just meant to be a quick way of constructing the tensor product of `S2gate `s  between modes `i` and `i+n`. It is just a lot faster to write the matrix directly than to compose it, especially since they all have the same squeezing parameter. It might be worth to fold this functionality inside the `fock_tensor` since it is only used there
2. You are right, the `vac` could be removed.

josh146 replied:
ahhh, maybe you could put `tensor_` in the name or something, makes it more clear 🙂

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+
+def choi_expand(S, r=np.arcsinh(1.0)):
+    r"""
+    Construct the Gaussian-Choi-Jamiolkowski expansion of a symplectic matrix S
```
Might be worth adding some mathematical details to the docstring body :)

```suggestion
    Construct the Gaussian-Choi-Jamiolkowski expansion of a symplectic matrix S.
```

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+
+def fock_tensor(S, alpha, cutoff, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of a Gaussian unitary represented by symplectic matrix S
```
```suggestion
    Calculates the Fock representation of a Gaussian unitary.
```

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+
+    (n, m) = S.shape
+    assert n == m
+    assert n % 2 == 0
```
better to make these both input validation (i.e., exceptions?)

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+    R = [1.0 / np.prod((np.tanh(r) ** i) / np.cosh(r)) for i in range(cutoff)]
+    # pylint: disable=assignment-from-no-return
+    lt = np.arctanh(np.linalg.svd(B, compute_uv=False))
+    T = np.exp(pref_exp) / (np.sqrt(np.prod(np.cosh(lt))))
```
Might be good to add line breaks and comments above, to explain the computation

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+    # There is probably a better way to do the following rescaling, but this is already "good"
+    for p in product(list(range(cutoff)), repeat=l):
+        tensor_view[p] = tensor_view[p] * np.prod([R[i] for i in p])
+    return tensor
```
Is the tensor returned using SF indexing?

nquesada replied:
No. But the easier can get into that format using `.transpose`

josh146 replied:
Do you think it's worth having a keyword argument for `sf_ind=True` to turn on Strawberry Fields indexing? I'm on the fence 🤔

josh146 commented on thewalrus/operations.py:
```diff
@@ -0,0 +1,120 @@
+    A = Amat(cov)
+    n, _ = A.shape
+    N = n // 2
+    B = A[0:N, 0:N].conj()
```
How come we take the conjugate of A here?

nquesada replied:
It is equivalent to taking the second half of the A matrix. More importantly it works!

josh146 commented on thewalrus/symplectic.py:
```diff
@@ -280,3 +281,18 @@ def rotation(theta):
+
+
+def sympmat(nmodes):
+    """Symplectic matrix of n modes
```
```suggestion
    """Symplectic matrix of :math:`n` modes.
```

josh146 commented on thewalrus/symplectic.py:
```diff
@@ -280,3 +281,18 @@ def rotation(theta):
+        nmodes (int): number of modes
+
+    Returns:
+        array: symplectic matrix of size 2*nmodes times 2*nmodes
```
```suggestion
        array: symplectic matrix of size :math:`2n\times2n`

josh146 commented on thewalrus/tests/test_operations.py:
```diff
@@ -0,0 +1,166 @@
+        ]
+    )
+    T = operations.fock_tensor(S, alphas, cutoff)
+    assert np.allclose(T, expected, atol=tol, rtol=0)
```
Should we also have a test for the S2 gate and beamsplitter?

nquesada replied:
For beamsplitter there are several, since interferometer supersedes beamsplitter.
For S2 it would be nice but not sure where to find reference values. There however selection rules that can be checked. Will add one for those rules.

josh146 commented on thewalrus/tests/test_operations.py:
```diff
@@ -0,0 +1,166 @@
+    U = np.sqrt(0.5) * np.array([[1, 1j], [1j, 1]])
+    alphas = np.zeros([nmodes])
+    S = interferometer(U)
+    T = operations.fock_tensor(S, alphas, cutoff)
```
if `operations` becomes `fock`, than might as well make `fock_tensor -> tensor`? So `fock.tensor(S, alphas, cutoff)`

josh146 commented on thewalrus/tests/test_operations.py:
```diff
@@ -0,0 +1,166 @@
+    for i, indi in enumerate(items):
+        for j, indj in enumerate(items):
+            mat[i, j] = T[tuple(indi + indj)]
+    np.allclose(U, mat[1 : 1 + nmodes, 1 : 1 + nmodes], atol=tol, rtol=0)
```
These two tests are a bit hard to follow, so might benefit from commenting :)

But great tests!

nquesada commented:
Hey @josh146   --- I commented more thoroughly the `operations.py` and rewrote most of the tests so that each test checks one and only one property. Hopefully it is more clear now. 
I was also thinking that it would actually make more sense to put this code inside `quantum.py`, since it is just a generalisation of the code to calculate the Fock representation of a pure state. What do you think?

Finally, thanks a lot for the code review. It was super useful!

nquesada commented on thewalrus/quantum.py:
```diff
@@ -534,7 +541,8 @@ def state_vector(mu, cov, post_select=None, normalize=False, cutoff=5, hbar=2, c
     alpha = beta[0:N]
     gamma = np.conj(alpha) - B @ alpha
-    pref = np.exp(-0.5 * (np.linalg.norm(alpha) ** 2 - alpha @ B @ alpha))
+    prefexp = -0.5 * (np.linalg.norm(alpha) ** 2 - alpha @ B @ alpha)
```
This line fixes a rather hard to catch bug!

nquesada commented:
OK. So I decided to move everything to `quantum.py`. The main function `fock_tensor` is in the end some matrix multiplications and the calling `state_vector` which is a lot easier to maintain. In the process of putting all these things together I found and corrected a very annoying bug in `state_vector`. Finally the tests that used to be in `test_operations` are now in `test_quantum.py`. I am now very happy with this state of this PR.

antalszava commented on thewalrus/quantum.py:
```diff
@@ -722,6 +731,30 @@ def is_classical_cov(cov, hbar=2, atol=1e-08):
+    if np.allclose(S.T @ Omega @ S, Omega, rtol=rtol, atol=atol):
+        return True
+
+    return False
```
This could also be simplified as:
```python
return np.allclose(S.T @ Omega @ S, Omega, rtol=rtol, atol=atol)
```

josh146 commented on thewalrus/quantum.py:
```diff
@@ -788,3 +821,77 @@ def total_photon_num_dist_pure_state(cov, cutoff=50, hbar=2, padding_factor=2):
+
+
+
+
```
Too many line breaks -- don't forget to `black -l 100 your/files.py` before merging!

nquesada replied:
Will do!

josh146 commented on thewalrus/quantum.py:
```diff
@@ -788,3 +821,77 @@ def total_photon_num_dist_pure_state(cov, cutoff=50, hbar=2, padding_factor=2):
+    x = 2 * alphat.real
+    p = 2 * alphat.imag
+    mu = np.concatenate([x, p])
+    tensor = state_vector(mu, cov, normalize=False, cutoff=cutoff, hbar=2, check_purity=False)
```
This is because `state_vector` uses `batch_hafnian` under the hood right?

nquesada replied:
correct!

josh146 commented on thewalrus/tests/test_quantum.py:
```diff
@@ -689,3 +696,198 @@ def test_total_photon_num_dist_pure_state(cutoff):
+        for q in product(list(range(cutoff)), repeat=nmodes):
+            if sum(p) != sum(q): #Check that there are the same total number of photons in the bra and the ket
+                r = tuple(list(p) + list(q))
+                np.allclose(T[r], 0.0, atol=tol, rtol=0)
```
```suggestion
                assert np.allclose(T[r], 0.0, atol=tol, rtol=0)
```

nquesada commented:
Alright. Thank @antalszava and @josh146 for your comments. I have included your suggestions in the latest version. I also moved `sympmat` and `is_symplectic` into `symplectic.py` which I think is their natural home. I also removed the function `choi_expand` and instead move its functionality inside `fock_tensor` . Finally, I also removed the function `TMS_cov` from `test_quantum.py` since the same functionality can be obtained by calling `two_mode_squeezing` from `symplectic.py`.
